Checks to pass critical only after reaching a failure threshold

PHBourquin opened the pull request:
A check may be set to become critical only if a specified number of successive
checks return critical in a row. Status will stay identical as before until
the threshold is reached.
This feature is available for HTTP, TCP, gRPC, Docker & Monitor checks.

PHBourquin committed:
Checks to passing/critical only after reaching a consecutive success/failure threshold

A check may be set to become passing/critical only if a specified number of successive
checks return passing/critical in a row. Status will stay identical as before until
the threshold is reached.
This feature is available for HTTP, TCP, gRPC, Docker & Monitor checks.

ShimmerGlass commented on agent/checks/check.go:
```diff
@@ -707,3 +690,33 @@ func (c *CheckGRPC) Stop() {
+		return
+	}
+
+	f.Logger.Printf("[WARN] agent: Check %q is critical but has not reached failure threshold %d/%d", checkID, f.failuresCounter, f.FailuresBeforeCritical)
```
A message in the style of "Check %q failed but has not reached failure threshold" would indicate more clearly that the check status is *not* critical.

ShimmerGlass reviewed (commented):
Looking good, a few things tho :

### Code
* The name `FailureHandler` does not reflect the structs job. Its methods are called when the check is passing too.
* I'd rather not have different names in the check structs : `FailureHandler` / `Notify`. They both satisfy the same interface.

### Docs
* There are already docs for the checks, you should update these
* The check types that support this new option should be documented

johncowen committed:
ui: Expands create-listeners API to allow recursive and functions (#5616)

1. Adds a Listeners class, which lets us...
2. Add Listeners recursively. So you can
createListeners().add(createListeners())
3. Also add the ability to `.add` as a plain function

This moves the entire idea more towards a generic teardown utility

johncowen committed:
ui: Reconcile ember-data store when records are deleted via blocking (#5745)

* ui: Reconciliate ember-data store when records are deleted via blocking

Currently we are barely using the ember-data store/cache, but it will
still cache records in the store even though technically we aren't using
it.

This adds a SyncTime to every record that uses blocking queries so we
can delete older records from the ember-data cache to prevent them
building up

* ui: Add basic timestamp method we can access from tests, fixup tests

Adds a timestamp method that we can access from within tests so we can
test that the SyncTime is being set.

There is probably a better way to do this, but this is also probably the
simplest approach - we are also likely to revisit this at a later date

johncowen committed:
ui: Allow text selection of clickable elements and their contents (#5770)

* ui: Allow text selection of clickable elements and their contents

This commit disables a click on mousedown be removing the `href`
attribute and moving it to a `data-href` attribute. On mouseup it will
only move it back if there is no selection. This means that an anchor
will only be followed on click _if_ there is no selection.

This fixes the fact that whenever you select some copy within a
clickable element it immediately throws you into the linked page when
you release your mouse.

Further notes:

We use the `isCollapsed` property here which 'seems' to be classed as
'experimental' in one place where I researched it:

https://developer.mozilla.org/en-US/docs/Web/API/Selection/isCollapsed

Although in others it makes no mention of this 'experimental' e.g:

- https://webplatform.github.io/docs/dom/Selection/isCollapsed/
- https://w3c.github.io/selection-api/#dom-selection-iscollapsed

I may have gone a little overboard in feature detection for this, but I
conscious of that fact that if `isCollapsed` doesn't exist at some point
in the future (something that seems unlikely). The code here will have
no effect on the UI. But I'd specifically like a second pair of eyes on
that.

* ui: Don't break right click, detects a secondary click on mousedown

* ui: Put anchor selection capability behind an ENV var

johncowen committed:
ui: Ensure an EventSource isn't opened if closed before first tick (#5703)

EventSources will wait for 1 tick before 'opening'. There is always the
chance that the EventSource is '.close()'ed before that tick. We
therefore check the 'readyState' before opening the EventSource

johncowen committed:
ui: Put some options in for event source runners (behind flags) (#5738)

johncowen committed:
ui: Adds ember-auto-import and replaces the 3 places where we used npm: (#5952)

Also removes ember-browserify

johncowen committed:
ui: CSS Components Migration (#5913)

Migrate roughly half of the base components into base
Adds a target for easily formatting CSS

Further CSS amends/migration (#5921)

1. tooltips within tables where a slightly bit troublesome due to a mix
of `inline-flex`, `overflow` and the need for truncation. This refineds
tooltips a slight bit more to work 'everywhere'.
2. We also move tooltip to use the correct color and min-width from
structure, but we overwrite the min-width here until we get confirmation
on widths/alignment of text within a tooltip.
3. Tiny fixes for breadcrumbs and toggle-buttons in tabular listings
4. Now we inline-flex our table cells, it means it is impossible to
truncate text without wrapping it in another element. This wraps all
Description like text in `<p>` tags. Generally the first column of text
is already wrapped in an `<a>` tag. Other items such as consul tags and
policy names etc get 'cutoff' rather than truncated.
5. We are now using all the icons from `@hashicorp/structure-icons`

johncowen committed:
ui: Include yadda/gherkin dictionaries we use for acc testing (#5956)

Currently these gherkin dictionaries are contained in
`ember-cli-api-double`, which isn't the correct place for them.
They belong with the steps themselves.
This is also less complicated now we can require/import javascript
using a standard approach

Also runs an update of `ember-cli-api-double` which will be the last
installation of version 1 of this module.

johncowen committed:
ui: Change vocab of ReopenableEventSource from 'reopen' to 'open' (#5973)

You can potentially close an EventSource before its first tick by
immediately setting the readyState to a non-open state. Therefore it
never opens.

Calling `open` will then open it.

'Open' fits better than 'reopen' when taking the above into account

johncowen committed:
ui: Reduce mutation of html.classList (#5974)

Throughout the app we mutate the value of the root node classList on
navigation between separate pages (basically on URL change).

Every template has a unique classList for example `template-service
template-show` and `template-service template-list` etc etc.

When navigating between 2 pages, both pages using the same template yet
with different data, previoulsy we would entirely clear out the
`html.classList` and then refill it again with eaxctly the same classes.

This commit moves this to perform a diff previous to mutating the
classList, and then potentially no classList mutating is needed when
moving between 2 pages of the same template.

johncowen committed:
ui: Enable creating listeners using an object/hash (#5975)

Makes listening to multiple events on one target slightly easier.
Adding events can be rolled up into passing through an object, and the
returned remove function removes all of the handlers in the object

For example:

```
//this.listen...
const remove = listeners.add(
  {
    'message': handler,
    'error': handler
  }
);
remove(); // removes all listeners in the object
```

The entire API for listeners is now becoming slightly overloaded, so
potentially we'd use this API always and remove the ability to use a
string/function pair.

johncowen committed:
ui: Small EventSource additions (#5978)

1. EventSources now pass themselves thorugh to the run function as a
second argument. This enables the use of arrow functions along with the
EventSources API
`(configuration, source) => source.close()`. Order of arguments could
potentially be switched at a later date.
2. BlockingEventSources now let you pass an 'event' through at
instantation time. If you do this, the event will immediately be dispatched
once the EventSource is opened. The usecase for this is for 'unfreezing' cached
BlockingEvents. This makes it easier to provide a cache for
BlockingEventSources by caching its data rather than the entire
BlockingEventSource itself.

```
new BlockingEventSource(
  (config, source) => { /* something */ },
  {
    cursor: 1024, // this would also come from a cache
    currentEvent: getFromSomeSortOfCache(eventSourceId) //this is the new bit
  }
);

// more realistically

new BlockingEventSource(
  (config, source) => { return data.findSomething(slug, config) },
  getFromSomeSortOfCache(eventSourceId)
);

```

johncowen committed:
ui: Adds the ability to frontend search instances by address (ip:port) (#5993)

johncowen committed:
ui: {{phrase-editor}} amends (#5991)

1. Re-focus the input element on phrase removal
2. Move all actions to `actions:`
3. Move to a form looking `value` rather than `items`
4. Move placeholder functionalit yinto the component
5. Force DDAU instead of two way binding with `slice` and `onchange`
6. Begin to deprecate the `searchable` interface

johncowen committed:
ui: removes `npm:` imports now we have standard importing (#6035)

johncowen committed:
ui: Ensure event sources are closed to the correct state when connecting (#6036)

This fails the tests, but wouldn't produce an actual bug as the source
is still set to closing, just 1 tick later.

johncowen committed:
ui: Be more exact with intent for table truncation (#6023)

Truncation should only happen currently on table content if the tag
within the td is the only child

johncowen committed:
ui: Add an additional 'more than 1 char' check for selection (#6022)

We'd rather 'play it safe' for our selection detection (which avoids
following links if you have some text selected). This commit adds an
additional check of 'more than 1 character' to the 'isSelected' check.
This is based on the fact that it's far quicker to type the character so
why would a user select on character? This therefore caters for any
fairly large (in the scheme of things) incidential mouse moves whilst
clicking a link.

Overall this is probably overkill

johncowen committed:
ui: Add `stats-card` component, integrate more `@hashicorp/structure-icons` (#6021)

1. Rebuild the heathchecked-resource component now we can copy and paste
2. As the above rebuild came with new icons, we also swapped out 'most'
of the other areas where we were using these new icons, plus any icons
that were effected by the new icon placeholders
3. Begin to remove more and more of the project specific icons (now
replaced by the shared ones)

johncowen committed:
ui: Show the correct message when a session has been removed from a KV (#6167)

Commit:
Run go test packages in parallel (#6165)

* modify gotestsum hacking that is unnecessary with the latest version

* try running 2 packages at a time

* try running 3 packages at a time

* remove old comments

Commit:
enable circleci go tests for forks and reorganize jobs (#6191)

kaitlincart committed:
[docs] New K8s-Consul deployment guide (#5859)

* New K8s-Consul deployment guide

* Update website/source/docs/guides/kubernetes-production-deploy.md

* Update website/source/docs/guides/kubernetes-production-deploy.md

Co-Authored-By: Rebecca Zanzig <rebecca@hashicorp.com>

* Update website/source/docs/guides/kubernetes-production-deploy.md

Co-Authored-By: Rebecca Zanzig <rebecca@hashicorp.com>

* Update website/source/docs/guides/kubernetes-production-deploy.md

Co-Authored-By: Rebecca Zanzig <rebecca@hashicorp.com>

* Update website/source/docs/guides/kubernetes-production-deploy.md

Co-Authored-By: Rebecca Zanzig <rebecca@hashicorp.com>

* Update website/source/docs/guides/kubernetes-production-deploy.md

Co-Authored-By: Rebecca Zanzig <rebecca@hashicorp.com>

* updating based on comments

* Update website/source/docs/guides/kubernetes-production-deploy.md

Co-Authored-By: Rebecca Zanzig <rebecca@hashicorp.com>

* Update website/source/docs/guides/kubernetes-production-deploy.md

* Update website/source/docs/guides/kubernetes-production-deploy.md

johncowen committed:
ui: Change layout styling for healthcheckoutput, add CheckID (#6195)

banks committed:
Allow raft TrailingLogs to be configured. (#6186)

This fixes pathological cases where the write throughput and snapshot size are both so large that more than 10k log entries are written in the time it takes to restore the snapshot from disk. In this case followers that restart can never catch up with leader replication again and enter a loop of constantly downloading a full snapshot and restoring it only to find that snapshot is already out of date and the leader has truncated its logs so a new snapshot is sent etc.

In general if you need to adjust this, you are probably abusing Consul for purposes outside its design envelope and should reconsider your usage to reduce data size and/or write volume.

banks committed:
Update CHANGELOG.md

mkeeler committed:
Update go-bexpr (#6190)

* Update go-bexpr to v0.1.1

This brings in:

• `in`/`not in` operators to do substring matching
• `matches` / `not matches` operators to perform regex string matching.

* Add the capability to auto-generate the filtering selector ops tables for our docs

mkeeler committed:
Update CHANGELOG.md

Commit:
remove travis config (#6199)

pierresouchay committed:
Display nicely Networks (CIDR) in runtime configuration (#6029)

* Display nicely Networks (CIDR) in runtime configuration

CIDR mask is displayed in binary in configuration.
This add support for nicely displaying CIDR in runtime configuration.

Currently, if a configuration contains the following lines:

  "http_config": {
    "allow_write_http_from": [
      "127.0.0.0/8",
      "::1/128"
    ]
  }

A call to `/v1/agent/self?pretty` would display

  "AllowWriteHTTPFrom": [
            {
                "IP": "127.0.0.0",
                "Mask": "/wAAAA=="
            },
            {
                "IP": "::1",
                "Mask": "/////////////////////w=="
            }
  ]

This PR fixes it and it will now display:

   "AllowWriteHTTPFrom": [ "127.0.0.0/8", "::1/128" ]

* Added test for cidr nice rendering in `TestSanitize()`.

mkeeler committed:
Update CHANGELOG.md

freddygv committed:
Update CHANGELOG.md

freddygv committed:
Restore NotifyListen to avoid panic in newServer retry (#6200)

Commit:
resolve circleci config conflicts

rboyer committed:
tests: adding new envoy integration tests for L7 service-resolvers (#6129)

Additionally:

- wait for bootstrap config entries to be applied

- run the verify container in the host's PID namespace so we can kill
envoys without mounting the docker socket

* assert that we actually send HEALTHY and UNHEALTHY endpoints down in EDS during failover

rboyer committed:
connect: rework how the service resolver subset OnlyPassing flag works (#6173)

The main change is that we no longer filter service instances by health,
preferring instead to render all results down into EDS endpoints in
envoy and merely label the endpoints as HEALTHY or UNHEALTHY.

When OnlyPassing is set to true we will force consul checks in a
'warning' state to render as UNHEALTHY in envoy.

Fixes #6171

rboyer committed:
update changelog

rboyer committed:
connect: validate and test more of the L7 config entries (#6156)

rboyer committed:
update changelog

rboyer committed:
connect: change router syntax for matching query parameters to resemble the syntax for matching paths and headers for consistency. (#6163)

This is a breaking change, but only in the context of the beta series.

rboyer committed:
connect: allow L7 routers to match on http methods (#6164)

Fixes #6158

rboyer committed:
update changelog

freddygv committed:
Make new config when retrying testServer creation (#6204)

Commit:
Fix links to ambassador website

mkeeler committed:
Envoy Mesh Gateway integration tests (#6187)

* Allow setting the mesh gateway mode for an upstream in config files

* Add envoy integration test for mesh gateways

This necessitated many supporting changes in most of the other test cases.

Add remote mode mesh gateways integration test

jefferai committed:
Chunking support (#6172)

* Initial chunk support

This uses the go-raft-middleware library to allow for chunked commits to the KV

mkeeler committed:
Update CHANGELOG.md

freddygv committed:
auto-encrypt: Fix port resolution and fallback to default port (#6205)

Auto-encrypt meant to fallback to the default port when it wasn't provided, but it hadn't been because of an issue with the error handling. We were checking against an incomplete error value:
"missing port in address" vs "address $HOST: missing port in address"

Additionally, all RPCs to AutoEncrypt.Sign were using a.config.ServerPort, so those were updated to use ports resolved by resolveAddrs, if they are available.

jefferai committed:
Make the chunking test multidimensional (#6212)

This ensures that it's not just a single operation we restores
successfully, but many. It's the same foundation, just with multiple
going on at once.

Commit:
add dev docker image upload (#5879)

* add dev docker image upload

* remove the go cache since that isn't needed

* add comment and image labels

* get password from stdin

Commit:
remove sdk and api packages (#6214)

Commit:
check vendor folder (#6006)

* check vendor in circleci

* make vendor checking step more descriptive

* reorganize check-vendor and gofmt jobs

mkeeler committed:
Allow forwarding of some status RPCs (#6198)

* Allow forwarding of some status RPCs

* Update docs

* add comments about not using the regular forward

mkeeler committed:
Update CHANGELOG.md

pearkes committed:
config: correct limit to limits in config example (#6219)

This isn't yet documented on the website, but wanted to update this to add the missing s.

mikemorris committed:
docs: add TCP half-close broken pipe to common errors (#6203)

pearkes committed:
Update CHANGELOG.md

hashicorp-ci committed:
update bindata_assetfs.go

hashicorp-ci committed:
Release v1.5.3

pearkes committed:
Putting source back into Dev Mode

rboyer committed:
Merge Consul OSS branch master at commit b3541c4f34d43ab92fe52256420759f17ea0ed73

rboyer committed:
Merge pull request #6223 from hashicorp/master-merge-b3541c4f3

Master merge b3541c4f3

Commit:
Merge pull request #6210 from hashicorp/docs/fix-ambassador-link

Fix links to ambassador website

mkeeler committed:
Set --max-obj-name-len 256 when execing Envoy (#6202)

* Pass -max-obj-name-len 256 to envoy

* Update test expectations.

* Add a note about requireing the max-obj-name-len option to be set

mkeeler committed:
Fix CA Replication when ACLs are enabled (#6201)

Secondary CA initialization steps are:

• Wait until the primary will be capable of signing intermediate certs. We use serf metadata to check the versions of servers in the primary which avoids needing a token like the previous implementation that used RPCs. We require at least one alive server in the primary and the all alive servers meet the version requirement.
• Initialize the secondary CA by getting the primary to sign an intermediate

When a primary dc is configured, if no existing CA is initialized and for whatever reason we cannot initialize a secondary CA the secondary DC will remain without a CA. As soon as it can it will initialize the secondary CA by pulling the primaries roots and getting the primary to sign an intermediate.

This also fixes a segfault that can happen during leadership revocation. There was a spot in the secondaryCARootsWatch that was getting the CA Provider and executing methods on it without nil checking. Under normal circumstances it wont be nil but during leadership revocation it gets nil'ed out. Therefore there is a period of time between closing the stop chan and when the go routine is actually stopped where it could read a nil provider and cause a segfault.

mkeeler committed:
Update CHANGELOG.md

Commit:
Merge remote-tracking branch 'origin/master' into release/1-6

hashicorp-ci committed:
update bindata_assetfs.go

hashicorp-ci committed:
Release v1.6.0-beta3

mkeeler committed:
Fix prepared query upstream endpoint generation (#6236)

Use the correct SNI value for prepared query upstreams

mkeeler committed:
Update CHANGELOG.md

mkeeler committed:
Update CHANGELOG.md

mkeeler committed:
Fix flaky tests (#6229)

hashicorp-ci committed:
Merge Consul OSS branch 'master' at commit a1725e6b5299c6ce12e8273205f90fba31403686

mkeeler committed:
Fix envoy canBind (#6238)

* Fix envoy cli canBind function

The string form of an Addr was including the CIDR causing the str equals to not match.

* Remove debug prints

mkeeler committed:
Update CHANGELOG.md

freddygv committed:
Update default gossip encryption key size to 32 bytes

freddygv committed:
Update CHANGELOG.md

Commit:
connect: Support RSA keys in addition to ECDSA (#6055)

Support RSA keys in addition to ECDSA

Commit:
connect: support AWS PCA as a CA provider (#6189)

Port AWS PCA provider from consul-ent

banks committed:
Revert "connect: support AWS PCA as a CA provider" (#6251)

This reverts commit 3497b7c00d49c4acbbf951d84f2bba93f3da7510.

s-mang committed:
fix 'consul connect envoy' to try to use previously-configured grpc port (#6245)

fix 'consul connect envoy' to try to use previously-configured grpc port on running agent before defaulting to 8502

Fixes #5011

venkatakrishnaannam committed:
docs: Fix minor mistakes in index.html.md (#6239)

omerzach committed:
Fix typo in architecture.html.md (#6261)

rboyer committed:
connect: validate upstreams and prevent duplicates (#6224)

* connect: validate upstreams and prevent duplicates

* Actually run Upstream.Validate() instead of ignoring it as dead code.

* Prevent two upstreams from declaring the same bind address and port.
  It wouldn't work anyway.

* Prevent two upstreams from being declared that use the same
  type+name+namespace+datacenter. Due to how the Upstream.Identity()
  function worked this ended up mostly being enforced in xDS at use-time,
  but it should be enforced more clearly at register-time.

rboyer committed:
update changelog

rboyer committed:
connect: reconcile how upstream configuration works with discovery chains (#6225)

* connect: reconcile how upstream configuration works with discovery chains

The following upstream config fields for connect sidecars sanely
integrate into discovery chain resolution:

- Destination Namespace/Datacenter: Compilation occurs locally but using
different default values for namespaces and datacenters. The xDS
clusters that are created are named as they normally would be.

- Mesh Gateway Mode (single upstream): If set this value overrides any
value computed for any resolver for the entire discovery chain. The xDS
clusters that are created may be named differently (see below).

- Mesh Gateway Mode (whole sidecar): If set this value overrides any
value computed for any resolver for the entire discovery chain. If this
is specifically overridden for a single upstream this value is ignored
in that case. The xDS clusters that are created may be named differently
(see below).

- Protocol (in opaque config): If set this value overrides the value
computed when evaluating the entire discovery chain. If the normal chain
would be TCP or if this override is set to TCP then the result is that
we explicitly disable L7 Routing and Splitting. The xDS clusters that
are created may be named differently (see below).

- Connect Timeout (in opaque config): If set this value overrides the
value for any resolver in the entire discovery chain. The xDS clusters
that are created may be named differently (see below).

If any of the above overrides affect the actual result of compiling the
discovery chain (i.e. "tcp" becomes "grpc" instead of being a no-op
override to "tcp") then the relevant parameters are hashed and provided
to the xDS layer as a prefix for use in naming the Clusters. This is to
ensure that if one Upstream discovery chain has no overrides and
tangentially needs a cluster named "api.default.XXX", and another
Upstream does have overrides for "api.default.XXX" that they won't
cross-pollinate against the operator's wishes.

Fixes #6159

rboyer committed:
update changelog

rboyer committed:
connect: simplify the compiled discovery chain data structures (#6242)

This should make them better for sending over RPC or the API.

Instead of a chain implemented explicitly like a linked list (nodes
holding pointers to other nodes) instead switch to a flat map of named
nodes with nodes linking other other nodes by name. The shipped
structure is just a map and a string to indicate which key to start
from.

Other changes:

* inline the compiler option InferDefaults as true

* introduce compiled target config to avoid needing to send back
  additional maps of Resolvers; future target-specific compiled state
  can go here

* move compiled MeshGateway out of the Resolver and into the
  TargetConfig where it makes more sense.

rboyer committed:
update changelog

rboyer committed:
server: if inserting bootstrap config entries fails don't silence the errors (#6256)

rboyer committed:
update changelog

johncowen committed:
ui: Upgrade ember-cli-api-double. Adds new staging env (#6136)

ember-cli-api-double has been upgraded for 3 things:

1. Use the correct configuration flags
2. Automatically include the necessary files to enable the api doubles
without requiring a server. This can be disabled to provide custom
functionality (so we can stitch our BDD style testing in with this)
3. When used statically, read the cookies from the users browser to
enable basic ad-hoc double editing (e.g. CONSUL_SERVICE_COUNT=100000)

Once upgraded we've now moved the config to the correct place, added a
new environment (staging) to use the static-style of doubles
The test environment continues to use custom cookie setting and url
checking so we disable this 'auto importing' by setting 'auto-import' to
false for the configuration for the addon.

We also added a couple of new package script targets to explicitly serve
or build the UI with the entirely static UI.

johncowen committed:
ui: [Bugfix] - Sticky KV Sessions (#6166)

Initialize session value to `null` to prevent stickiness from a session the previous KV

johncowen committed:
ui: Adds readonly meta data to the serviceInstance and node detail pages (#6196)

rboyer committed:
connect: detect and prevent circular discovery chain references (#6246)

rboyer committed:
update changelog

alvin-huang committed:
Add arm builds (#6263)

* try arm builds

* Update .circleci/config.yml

Co-Authored-By: Matt Keeler <mkeeler@users.noreply.github.com>

* Update .circleci/config.yml

Co-Authored-By: Matt Keeler <mkeeler@users.noreply.github.com>

* Update .circleci/config.yml

Co-Authored-By: Matt Keeler <mkeeler@users.noreply.github.com>

Commit:
connect: generate intermediate at same time as root (#6272)

Generate intermediate at same time as root
Co-Authored-By: Freddy <freddygv@users.noreply.github.com>

alvin-huang committed:
add generic master merge into release/* branches (#6249)

mkeeler committed:
Add license management functions to API client (#6268)

* Add license management functions to API client

* Get rid of jsonapi struct tags

mkeeler committed:
Update CHANGELOG.md

rboyer committed:
connect: expose an API endpoint to compile the discovery chain (#6248)

In addition to exposing compilation over the API cleaned up the structures that would be exchanged to be cleaner and easier to support and understand.

Also removed ability to configure the envoy OverprovisioningFactor.

rboyer committed:
update changelog

alvin-huang committed:
Merge remote-tracking branch 'origin/master' into release/1-6

alvin-huang committed:
fix grpc-addr-config hosts template

alvin-huang committed:
Merge pull request #6274 from hashicorp/merge-master-de01a1e

Merge master at de01a1e279230624fcc2d7e692b7e773d570204b

johncowen committed:
ui: Enable blocking queries by default (#6194)

-Enable blocking queries by default
-Change assertion to check for the last PUT request, not just any request for session destruction from a node page.

Since we've now turned on blocking queries by default this means that a
second GET request is made after the PUT request that we are asserting
for but before the assertion itself, this meant the assertion failed. We
double checked this by turning off blocking queries for this test using

```
And settings from yaml
---
consul:client:
  blocking: 0
---
```

which made the test pass again.

As moving forwards blocking queries will be on by default, we didn't
want to disable blocking queries for this test, so we now assert the
last PUT request specifically. This means we continue to assert that the
session has been destroyed but means we don't get into problems of
ordering of requests here

johncowen committed:
ui: add some additional env level URLs for easier linking (#6145)

Right now we only use CONSUL_DOCS_URL, the others are for future usage

rboyer committed:
connect: fix failover through a mesh gateway to a remote datacenter (#6259)

Failover is pushed entirely down to the data plane by creating envoy
clusters and putting each successive destination in a different load
assignment priority band. For example this shows that normally requests
go to 1.2.3.4:8080 but when that fails they go to 6.7.8.9:8080:

- name: foo
  load_assignment:
    cluster_name: foo
    policy:
      overprovisioning_factor: 100000
    endpoints:
    - priority: 0
      lb_endpoints:
      - endpoint:
          address:
            socket_address:
              address: 1.2.3.4
              port_value: 8080
    - priority: 1
      lb_endpoints:
      - endpoint:
          address:
            socket_address:
              address: 6.7.8.9
              port_value: 8080

Mesh gateways route requests based solely on the SNI header tacked onto
the TLS layer. Envoy currently only lets you configure the outbound SNI
header at the cluster layer.

If you try to failover through a mesh gateway you ideally would
configure the SNI value per endpoint, but that's not possible in envoy
today.

This PR introduces a simpler way around the problem for now:

1. We identify any target of failover that will use mesh gateway mode local or
   remote and then further isolate any resolver node in the compiled discovery
   chain that has a failover destination set to one of those targets.

2. For each of these resolvers we will perform a small measurement of
   comparative healths of the endpoints that come back from the health API for the
   set of primary target and serial failover targets. We walk the list of targets
   in order and if any endpoint is healthy we return that target, otherwise we
   move on to the next target.

3. The CDS and EDS endpoints both perform the measurements in (2) for the
   affected resolver nodes.

4. For CDS this measurement selects which TLS SNI field to use for the cluster
   (note the cluster is always going to be named for the primary target)

5. For EDS this measurement selects which set of endpoints will populate the
   cluster. Priority tiered failover is ignored.

One of the big downsides to this approach to failover is that the failover
detection and correction is going to be controlled by consul rather than
deferring that entirely to the data plane as with the prior version. This also
means that we are bound to only failover using official health signals and
cannot make use of data plane signals like outlier detection to affect
failover.

In this specific scenario the lack of data plane signals is ok because the
effectiveness is already muted by the fact that the ultimate destination
endpoints will have their data plane signals scrambled when they pass through
the mesh gateway wrapper anyway so we're not losing much.

Another related fix is that we now use the endpoint health from the
underlying service, not the health of the gateway (regardless of
failover mode).

rboyer committed:
Update CHANGELOG.md

s-mang committed:
fallback to proxy config global protocol when upstream services' protocol is unset (#6277)

fallback to proxy config global protocol when upstream services' protocol is unset

Fixes #5857

rboyer committed:
api: un-deprecate api.DecodeConfigEntry (#6278)

Add clarifying commentary about when it is not safe to use it. Also add
tests.

rboyer committed:
update changelog

hashicorp-ci committed:
Merge Consul OSS branch 'master' at commit d84863799deca45ccf4bec5ab9f645ccae6b3aeb

rboyer committed:
connect: ensure intention replication continues to work when the replication ACL token changes (#6288)

rboyer committed:
update changelog

mkeeler committed:
mesh-gateway ACL tokens should also have `node:read` on everyth… (#6291)

rboyer committed:
docs: add documentation for discovery chains

Fixes #6273

rboyer committed:
command: ensure that the json form of config entries can be submitted with 'consul config write' (#6290)

The json decoder inside of the HCLv1 hcl.Decode function behaves
unexpectedly when decoding generically into a map[string]interface{} as
is done for 'consul config write' pre-submit decoding.

This results in some subtle (service-router Match and Destinations being
separated) and some not so subtle (service-resolver subsets and failover
panic if multiple subsets are referenced) bugs when subsequently passed
through mapstructure to finish decoding.

Given that HCLv1 is basically frozen and the HCL part of it is fine
instead of trying to figure out what the underlying bug is in the json
decoder for our purposes just sniff the byte slice and selectively use
the stdlib json decoder for JSON and hcl decoder for HCL.

rboyer committed:
update changelog

mikemorris committed:
connect: remove managed proxies (#6220)

* connect: remove managed proxies implementation and all supporting config options and structs

* connect: remove deprecated ProxyDestination

* command: remove CONNECT_PROXY_TOKEN env var

* agent: remove entire proxyprocess proxy manager

* test: remove all managed proxy tests

* test: remove irrelevant managed proxy note from TestService_ServerTLSConfig

* test: update ContentHash to reflect managed proxy removal

* test: remove deprecated ProxyDestination test

* telemetry: remove managed proxy note

* http: remove /v1/agent/connect/proxy endpoint

* ci: remove deprecated test exclusion

* website: update managed proxies deprecation page to note removal

* website: remove managed proxy configuration API docs

* website: remove managed proxy note from built-in proxy config

* website: add note on removing proxy subdirectory of data_dir

mikemorris committed:
website: restore accidental JSON deletion [skip ci] (#6303)

rboyer committed:
connect: allow mesh gateways to use central config (#6302)

rboyer committed:
update changelog

grove-mountain committed:
docs: Update consul-helm example to pull latest tag

mkeeler committed:
Update the v1/agent/service/:service endpoint to output tagged… (#6304)

mkeeler committed:
Update CHANGELOG.md

hashicorp-ci committed:
Merge Consul OSS branch 'master' at commit 8241787e922955e973c0e762ad3cb8db1804f6cd

johncowen committed:
ui: fix/replace magnifier icon (#6308)

johncowen committed:
ui: Leader icon for node listing view (#6265)

- yarn upgrade consul-api-double which includes `status/leader`
- add all the ember-data things required to call a new endpoint
- Pass the new leader variable through to the template
- use the new leader variable in the template to set a leader
- add acceptance testing to verify leaders are highlighted
- Change testing navigation/api requests to status/leader (on the node listing page, status/leader is now the last get request to
be called).
- Template whitespace commit (less indenting)
- adds a test to to assert no errors happen with an unelected leader

mikemorris committed:
snapshot: add TLS support to HalfCloser interface (#6216)

Calls net.TCPConn.CloseWrite or mtls.Conn.CloseWrite, which was added in https://go-review.googlesource.com/c/go/+/31318/

s-mang committed:
add flag to allow /operator/keyring requests to only hit local servers (#6279)

Add parameter local-only to operator keyring list requests to force queries to only hit local servers (no WAN traffic).

HTTP API: GET /operator/keyring?local-only=true
CLI: consul keyring -list --local-only

Sending the local-only flag with any non-GET/list request will result in an error.

mkeeler committed:
Add missing LicenseReset API function (#6311)

johncowen committed:
ui: Adds a `sort-control` component for asc/desc sorting of columns etc (#6034)

This adds the component but doesn't yet use it anywhere. No tests
are added here as there isn't an awful lot to test.

s-mang committed:
Update CHANGELOG.md

mikemorris committed:
changelog: add managed proxy removal to breaking changes

mikemorris committed:
changelog: add snapshot half-close fix

hashicorp-ci committed:
Merge Consul OSS branch 'master' at commit 8f7586b339dbb518eff3a2eec27d7b8eae7a3fbb

mkeeler committed:
Fix changelog format (and alphabatize stuff)

hashicorp-ci committed:
update bindata_assetfs.go

hashicorp-ci committed:
Release v1.6.0-rc1

mkeeler committed:
Update toolchain to Go 1.12.8 (#6319)

rboyer committed:
agent: blocking central config RPCs iterations should not interfere with each other (#6316)

rboyer committed:
update changelog

mkeeler committed:
Update CHANGELOG.md

mkeeler committed:
Update to google.golang.org/grpc v1.23.0 (#6320)

mkeeler committed:
Update CHANGELOG.md

mattc41190 committed:
Fix 404 (#6300)

On page: https://www.consul.io/discovery.html

If you click the link for Health Checks -> Learn More the underlying resource is:

https://learn.hashicorp.com/consul/getting-started/checks

This page for me is a 404. I think you've bundled it together in the following page:

Register a Service and Health Check - Service Discovery

Located at: https://learn.hashicorp.com/consul/getting-started/services

Thanks for Consul, it's really awesome.

hashicorp-ci committed:
Merge Consul OSS branch 'master' at commit 23cf22960af00a02530f51327d9566c8b52e4de7

rboyer committed:
xds: improve how envoy metrics are emitted (#6312)

Since generated envoy clusters all are named using (mostly) SNI syntax
we can have envoy read the various fields out of that structure and emit
it as stats labels to the various telemetry backends.

I changed the delimiter for the 'customization hash' from ':' to '~'
because ':' is always reencoded by envoy as '_' when generating metrics
keys.

rboyer committed:
update changelog

mkeeler committed:
Filter out left/leaving serf members when determining if new AC… (#6332)

mkeeler committed:
Update CHANGELOG.md

tryan225 committed:
Clarifying autopilot bootstrap and config options

pearkes committed:
website: update the vs. envoy and proxies page (#6326)

* website: update the vs. envoy and proxies page

This is the second result on Google for "consul envoy" and
it seemed like it needed a bit of an upgrade to help clarify the
current state.

* Update website/source/intro/vs/proxies.html.md

Co-Authored-By: Judith Malnick <judith.patudith@gmail.com>

* Update website/source/intro/vs/proxies.html.md

Co-Authored-By: Judith Malnick <judith.patudith@gmail.com>

* Update website/source/intro/vs/proxies.html.md

Co-Authored-By: Judith Malnick <judith.patudith@gmail.com>

* Update website/source/intro/vs/proxies.html.md

Co-Authored-By: Judith Malnick <judith.patudith@gmail.com>

* Apply suggestions from code review

Co-Authored-By: Judith Malnick <judith.patudith@gmail.com>

rboyer committed:
connect: updating a service-defaults config entry should leave an unset protocol alone (#6342)

If the entry is updated for reasons other than protocol it is surprising
that the value is explicitly persisted as 'tcp' rather than leaving it
empty and letting it fall back dynamically on the proxy-defaults value.

rboyer committed:
update changelog

rboyer committed:
connect: introduce ExternalSNI field on service-defaults (#6324)

Compiling this will set an optional SNI field on each DiscoveryTarget.
When set this value should be used for TLS connections to the instances
of the target. If not set the default should be used.

Setting ExternalSNI will disable mesh gateway use for that target. It also 
disables several service-resolver features that do not make sense for an 
external service.

rboyer committed:
connect: generate the full SNI names for discovery targets in the compiler rather than in the xds package (#6340)

rboyer committed:
update changelog

rboyer committed:
connect: ensure time.Duration fields retain their human readable forms in the API (#6348)

This applies for both config entries and the compiled discovery chain.

Also omit some other config entries fields when empty.

rboyer committed:
update changelog

tryan225 committed:
Merge pull request #6341 from tryan225/docs/autopilot-updates

Clarifying autopilot bootstrap and config options

johncowen committed:
docs: Fix typo layey > layer (#6352)

mkeeler committed:
Turned on Envoy 1.11.1 integration tests (#6347)

I also ran this against 1.5.2 so the docs update claiming compatibility should still be accurate.

mkeeler committed:
Ensure that config entry writes are forwarded to the primary DC (#6339)

mkeeler committed:
Update CHANGELOG.md

hashicorp-ci committed:
Merge Consul OSS branch 'master' at commit a7ded1bd8efcbc3c67978f050b6f16ec5e8a832d

famousgarkin committed:
docs: ports docs missing paren fix (#6367)

rboyer committed:
docs: fixing L7 config entries documentation (#6358)

- add service-router example involving gRPC
- fix indentation on service-router page by splitting it up
- remove reference to removed setting

alvin-huang committed:
remove ci merge branches if nightly merge fails (#6369)

alvin-huang committed:
Merge Consul OSS branch 'master' at commit ce9cfc773d529ae4b8259d195323c0c350c1f9f1

rboyer committed:
docs: document how envoy escape hatches work with the discovery chain (#6350)

- Bootstrap escape hatches are OK.
- Public listener/cluster escape hatches are OK.
- Upstream listener/cluster escape hatches are not supported.

If an unsupported escape hatch is configured and the discovery chain is
activated log a warning and act like it was not configured.

Fixes #6160

rboyer committed:
docs: remove beta references; leave version notation (#6372)

pearkes committed:
website: fix typo on mesh page (#6368)

Fixes #6345.

alvin-huang committed:
print a proper error message for master merge CI job failures (#6371)

alvin-huang committed:
Merge Consul OSS branch 'master' at commit 59150281c52fafa4a3320c42128f1caf9dbd083b

rboyer committed:
connect: allow 'envoy_cluster_json' escape hatch to continue to function (#6378)

rboyer committed:
update changelog

hanshasselberg committed:
auto_encrypt: use server-port (#6287)

AutoEncrypt needs the server-port because it wants to talk via RPC. Information from gossip might not be available at that point and thats why the server-port is being used.

danielehc committed:
Update agent.html.markdown.erb (#6380)

Adding a note on how to make Consul trust S3-compatible storage that expose a self-signed certificate.

arj22 committed:
Update observability.html.md (#6379)

mkeeler committed:
Secondary CA `establishLeadership` fix (#6383)

This prevents ACL issues (or other issues) during intermediate CA cert signing from failing leader establishment.

rboyer committed:
Merge pull request #6388 from hashicorp/release/1-6

merging release/1-6 into master

hanshasselberg committed:
Update CHANGELOG.md

mkeeler committed:
Update CHANGELOG.md

rboyer committed:
flatten 1.6 series changelog (#6390)

rboyer committed:
api: bump to sdk v0.2.0

rboyer committed:
mod: bump main module to api v1.2.0 and sdk v0.2.0

hashicorp-ci committed:
update bindata_assetfs.go

hashicorp-ci committed:
Release v1.6.0

hanshasselberg committed:
make sure auto_encrypt has private key type and bits

hanshasselberg committed:
Update CHANGELOG.md

hanshasselberg committed:
Update CHANGELOG.md

Co-Authored-By: Freddy <freddygv@users.noreply.github.com>

hanshasselberg committed:
Update CHANGELOG.md

Co-Authored-By: Alvin Huang <17609145+alvin-huang@users.noreply.github.com>

rboyer committed:
Putting source back into Dev Mode

freddygv committed:
Rephrase bind docs (#6394)

abennett committed:
update dependencies for enterprise change (#6395)

rboyer committed:
test: don't leak agent goroutines in TestAgent_sidecarServiceFromNodeService (#6396)

A goroutine dump using runtime.Stack() before/after shows a drop from 121 => 4.

rboyer committed:
test: actually run envoy 1.11.1 integration tests on PRs (#6397)

mikemorris committed:
bump eventmachine to 1.2.7 in Gemfile.lock (#6389)

hanshasselberg committed:
auto_encrypt: verify_incoming_rpc is good enough for auto_encrypt.allow_tls (#6376)

Previously `verify_incoming` was required when turning on `auto_encrypt.allow_tls`, but that doesn't work together with HTTPS UI in some scenarios. Adding `verify_incoming_rpc` to the allowed configurations.

hanshasselberg committed:
make sure auto_encrypt has private key type and bits (#6392)

hanshasselberg committed:
changelog: add known issues section for 1.6.

Update CHANGELOG.md

nfagerlund committed:
website: Update middleman-hashicorp container and Gemfile.lock (#6374)

* website: Update middleman-hashicorp container and Gemfile.lock

Time marches on, and so do security vulnerabilities in Nokogiri. So it's time
for a new container.

As with last time, here's a reminder for the next person who needs to update
this:

- You shouldn't just update the dependency in Gemfile.lock, because your build
  times will go to heck as you compile Nokogiri from source on every run. So you
  need an updated container with all the dependencies.
- To update the container, you need to push a new tag to the middleman-hashicorp
  repo. Teamcity does the rest, and will ship a new container to Docker Hub
  (unless its credentials are out of date, in which case go ask team-eng-serv.)
- Once that's pushed:
    - Update Makefile
    - Update the Gemfile
    - Delete Gemfile.lock
    - `make website` until it comes up, then ctrl-C
    - Commit the changes

* website: Specify a different json version in Gemfile.lock

The Consul website uses different containers for preview and deploy, and this
oddball JSON version was causing issues. This commit sacrifices a little bit
of preview startup speed for (hopefully) working deploys.

hanshasselberg committed:
Update CHANGELOG.md

alvin-huang committed:
add nil pointer check for pointer to ACLToken struct (#6407)

rboyer committed:
test: fix TestAgent.Start() to not segfault if the DNSServer cannot ListenAndServe (#6409)

The embedded `Server` field on a `DNSServer` is only set inside of the
`ListenAndServe` method. If that method fails for reasons like the
address being in use and is not bindable, then the `Server` field will
not be set and the overall `Agent.Start()` will fail.

This will trigger the inner loop of `TestAgent.Start()` to invoke
`ShutdownEndpoints` which will attempt to pretty print the DNS servers
using fields on that inner `Server` field. Because it was never set,
this causes a nil pointer dereference and crashes the test.

rboyer committed:
test: send testagent logs through testing.Logf (#6411)

tradel committed:
Added DC and domain args to Configure method

tradel committed:
add domain and nodeName to agent cert request

tradel committed:
add serviceID to leaf cert request

tradel committed:
construct a common name for each CSR

tradel committed:
add subject names to issued certs

tradel committed:
create a common name for autoTLS agent certs

tradel committed:
confi\gure providers with DC and domain

tradel committed:
update tests to match new method signatures

alvin-huang committed:
revert commits on master (#6413)

pierresouchay committed:
[BUGFIX][BUILD] When test fail in circle-ci in main, have a proper error message (#6416)

Since FUNCNAME is not defined when running outside a function,
trap does not work and display wrong error message.

Example from https://circleci.com/gh/hashicorp/consul/69506 :

```
  ⨯ FAIL
  /home/circleci/project/test/integration/connect/envoy/run-tests.sh: line 1: FUNCNAME[0]: unbound variable
  make: *** [GNUmakefile:363: test-envoy-integ] Error 1
```

This fix will avoid this error message and display the real cause.

johncowen committed:
ui: Netlify previews (#6419)

- Add netlify redirects file
- Add netlify makefile target and commented config settings

pierresouchay committed:
Display IPs of machines when node names conflict to ease troubleshooting

When there is an node name conflicts, such messages are displayed within Consul:

`consul.fsm: EnsureRegistration failed: failed inserting node: Error while renaming Node ID: "e1d456bc-f72d-98e5-ebb3-26ae80d785cf": Node name node001 is reserved by node 05f10209-1b9c-b90c-e3e2-059e64556d4a with name node001`

While it is easy to find the node that has reserved the name, it is hard to find
the node trying to aquire the name since it is not registered, because it
is not part of `consul members` output

This PR will display the IP of the offender and solve far more easily those issues.

mkeeler committed:
TxnCheckOp has a `Check` field not a `Service` field (#6418)

rboyer committed:
test: add additional http status code assertions in coordinate HTTP API tests (#6410)

When this test flakes sometimes this happens:

    --- FAIL: TestCoordinate_Node (1.69s)
    panic: interface conversion: interface {} is nil, not structs.Coordinates [recovered]
    FAIL    github.com/hashicorp/consul/agent       19.999s
    Exit code: 1
    panic: interface conversion: interface {} is nil, not structs.Coordinates [recovered]
            panic: interface conversion: interface {} is nil, not structs.Coordinates

There is definitely a bug lurking, but the code seems to imply this can
only return nil on 404. The tests previously were not checking the
status code.

The underlying cause of the flake is unknown, but this should turn the
failure into a more normal test failure.

rboyer committed:
test: explicitly run the pprof tests for 1s instead of the 30s default (#6421)

rboyer committed:
test: ensure the node name is a valid dns name (#6424)

The space in the node name was making every test emit a useless warning.

mkeeler committed:
Store primaries root in secondary after intermediate signature (#6333)

* Store primaries root in secondary after intermediate signature

This ensures that the intermediate exists within the CA root stored in raft and not just in the CA provider state. This has the very nice benefit of actually outputting the intermediate cert within the ca roots HTTP/RPC endpoints.

This change means that if signing the intermediate fails it will not set the root within raft. So far I have not come up with a reason why that is bad. The secondary CA roots watch will pull the root again and go through all the motions. So as soon as getting an intermediate CA works the root will get set.

* Make TestAgentAntiEntropy_Check_DeferSync less flaky

I am not sure this is the full fix but it seems to help for me.

mkeeler committed:
Update CHANGELOG.md

s-mang committed:
txn: don't try to decode request bodies > raft.SuggestedMaxDataSize (#6422)

txn: don't try to decode request bodies > raft.SuggestedMaxDataSize

Commit:
Add option to register services and their checks idempotently (#4905)

freddygv committed:
Update CHANGELOG.md

freddygv committed:
Update CHANGELOG.md

pierresouchay committed:
Distinguish between DC not existing and not being available (#6399)

s-mang committed:
refactor & add better retry logic to NewTestAgent (#6363)

Fixes #6361

johncowen committed:
ui: Fix metaData test as result of new consul-api-double (#6436)

johncowen committed:
ui: re-remove the `Type` header on the service listing page (#6439)

The `Type` column used for giving details on what type of a service each
item is was removed in https://github.com/hashicorp/consul/pull/6075.

As a result of keeping long running branches in sync, this change was
partly reverted in an earlier PR (the type header was re-added)
https://github.com/hashicorp/consul/pull/5913 following a rebase.

This commit re-removes the `Type` table header (the `<th>`)

johncowen committed:
Merge pull request #6437 from hashicorp/ui-staging

ui: UI Release Merge (ui-staging merge)

johncowen committed:
Update CHANGELOG.md

s-mang committed:
remove funky panic/recover in agent tests (#6442)

johncowen committed:
docs: Fix discovery chain internals link (#6449)

s-mang committed:
test: ensure all TestAgent constructions use a constructor (#6443)

ensure all TestAgent constructions use a constructor to get start retries + test logs going to the right place

Fixes #6435

alvin-huang committed:
change DOCKER_LOGIN env name to DOCKER_USER (#6451)

rboyer committed:
test: actually wait for the TestAgent to be fully shutdown (#6441)

dcallao committed:
docs: added Consul Integration Program Docs Page (#6459)

* adding the Consul Integration Program page in docs section

dcallao committed:
fixed broken links on consul program page (#6463)

judwhite committed:
docs: fix typo in install/performance (#6428)

hanshasselberg committed:
agent: handleEnterpriseLeave (#6453)

hanshasselberg committed:
Update CHANGELOG.md

rboyer committed:
ci: inject missing tar binary (#6477)
